test/src/test/java/jenkins/telemetry/MissingClassTelemetryTest.java

```diff
@@ -0,0 +1,147 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.telemetry;
+
+import hudson.ExtensionList;
+import hudson.model.UnprotectedRootAction;
+import hudson.security.csrf.CrumbExclusion;
+import jenkins.telemetry.impl.java11.CatcherClassLoader;
+import jenkins.telemetry.impl.java11.MissingClassTelemetry;
+import net.sf.json.JSONArray;
+import net.sf.json.JSONObject;
+import org.apache.commons.io.IOUtils;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.TestExtension;
+import org.kohsuke.stapler.StaplerRequest;
+import org.kohsuke.stapler.StaplerResponse;
+
+import javax.annotation.CheckForNull;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.StringWriter;
+import java.nio.charset.StandardCharsets;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * This test needs to be here to be able to modify the {@link Telemetry#ENDPOINT} as it's package protected.
+ */
+public class MissingClassTelemetryTest {
+    private static final String TELEMETRY_ENDPOINT = "uplink";
+    private CatcherClassLoader cl;
+
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    private static JSONObject received = null;
+
+    @Before
+    public void prepare() throws Exception {
+        received = null;
+        cl = new CatcherClassLoader(this.getClass().getClassLoader());
+        Telemetry.ENDPOINT = j.getURL().toString() + TELEMETRY_ENDPOINT + "/events";
+        j.jenkins.setNoUsageStatistics(false); // tests usually don't submit this, but we need this
+    }
+
+    /**
+     * Test if the telemetry sent works and the received data is the expected for a specific case (5 occurrences of the
+     * same stack trace).
+     * @throws InterruptedException if the thread is interrupted while sleeping
+     */
+    @Test
+    public void telemetrySentWorks() throws InterruptedException {
+        // Generate 5 events
+        for(int i = 0; i < 5; i++) {
+            try {
+                cl.loadClass("sun.java.MyNonExistentClass");
+            } catch (ClassNotFoundException ignored) {
+            }
+        }
+
+        // Run the telemetry sent
+        ExtensionList.lookupSingleton(Telemetry.TelemetryReporter.class).doRun();
+        do {
+
+            Thread.sleep(250);
```

Comment:
You're killing me here! In a test!

Reply:
So glad I dodged a review from you in #3604 😄

Reply:
Yeah, :-D my mentor is Walt Disney (C&P): https://www.youtube.com/watch?v=vh84g8rC2oA

Reply:
Anyway, the test is not relying on sleep. It's not affected by the environment slowness. The real semaphore is `} while (received == null)`

Reply:
A timeout on a wait for a semaphore would work better in a CI environment.